### PR TITLE
MM-28226: Admin Advisor: add mitigations

### PR DESCRIPTION
**Pull request by catalintomai**
#### Summary
Add various mitigations for Admin Advisor metric warning job.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-28226

#### Related Pull Requests
https://github.com/mattermost/mattermost-webapp/pull/6353

### CODE REVIEW

File: app/server.go [hunk 1]
```diff
@@ -1201,21 +1201,22 @@ func doCheckNumberOfActiveUsersWarnMetricStatus(a *App) {
...
 		}
 
-		if nErr = a.Srv().Store.System().SaveOrUpdate(&model.System{Name: warnMetric.Id, Value: model.WARN_METRIC_STATUS_LIMIT_REACHED}); nErr != nil {
-			mlog.Error("Unable to write to database.", mlog.String("id", warnMetric.Id), mlog.Err(nErr))
-			continue
-		}
 		warnMetricStatus, _ := a.getWarnMetricStatusAndDisplayTextsForId(warnMetric.Id, nil)
-
 		if !warnMetric.IsBotOnly {
+			//banner and bot metrics - send websocket event
```
Comment (fmunshi) [inline app/server.go:1211]: nit space at start of comment 
```suggestion
			// Banner and bot metrics - send websocket event
```

File: app/server.go [hunk 2]
```diff
@@ -1201,21 +1201,22 @@ func doCheckNumberOfActiveUsersWarnMetricStatus(a *App) {
...
 			message := model.NewWebSocketEvent(model.WEBSOCKET_WARN_METRIC_STATUS_RECEIVED, "", "", "", nil)
 			message.Add("warnMetricStatus", warnMetricStatus.ToJson())
 			a.Publish(message)
+
+			//bot and banner metrics, do not send the bot message again
```
Comment (fmunshi) [inline app/server.go:1216]:
```suggestion
			// Bot and banner metrics, do not send the bot message again
```

File: i18n/en.json [hunk 3]
```diff
@@ -1996,39 +1996,39 @@
...
+    "translation": "Your Mattermost system now has 200 users. As your user base grows, provisioning new accounts can become time-consuming. We recommend that you integrate your organization’s Active Directory/LDAP, which will allow anyone with an account to access Mattermost.\r\n\r\n[Learn more about integrating with AD/LDAP](https://docs.mattermost.com/deployment/sso-ldap.html?utm_medium=product&utm_source=mattermost-advisor-bot&utm_content=adldap)\r\n\r\nBy clicking Acknowledge, you'll be sharing your information with Mattermost Inc., to learn more about upgrading. [Learn more](https://mattermost.com/pl/default-admin-advisory)"
   },
   {
     "id": "api.server.warn_metric.number_of_active_users_200.notification_title",
     "translation": "Integrate AD/LDAP"
```
Comment (srkgupta) [inline i18n/en.json:2015]: The Header of the bot message should also be updated as per the screenshot attached in the ticket https://mattermost.atlassian.net/browse/MM-28226.

File: i18n/en.json [hunk 4]
```diff
@@ -1996,39 +1996,39 @@
...
+    "translation": "Your Mattermost system now has 400 users. When you connect Mattermost with your organization's single sign-on provider, users can access Mattermost without having to re-enter their credentials. We recommend you integrate SAML 2.0 provider with your Mattermost server.\r\n\r\n[Learn more about integrating with SAML 2.0](https://docs.mattermost.com/deployment/sso-saml.html?utm_medium=product&utm_source=mattermost-advisor-bot&utm_content=saml)\r\n\r\nBy clicking Acknowledge, you'll be sharing your information with Mattermost Inc., to learn more about upgrading. [Learn more](https://mattermost.com/pl/default-admin-advisory)"
   },
   {
     "id": "api.server.warn_metric.number_of_active_users_400.notification_title",
     "translation": "Integrate SAML 2.0"
```
Comment (srkgupta) [inline i18n/en.json:2023]: The Header of the bot message should also be updated as per the screenshot attached in the ticket https://mattermost.atlassian.net/browse/MM-28226.

**Comment by catalintomai** [comment]
> doCheckNumberOfActiveUsersWarnMetricStatus

Agree, I couldn't find a good place for job business logic tests though - server_test does not cover that. Will look into it one more time.

**Review by srkgupta — CHANGES_REQUESTED** [review]
Hi @catalintomai 

I am seeing the following issues:

1. Header is not changed for 200-user and 400-user advisory. As per the JIRA ticket, it is expected to change the header for these advisory messages as well. 

2. Banner does not reappear when the nag interval has passed. As per the expectations mentioned in the JIRA ticket, 

`Make banner not reappear upon page refresh (it should only reappear when another advisory is triggered or when the nag interval has passed)`

Infact, for me even when I removed the warn_metric_number_of_active_users_500 entry from systems table, the banner never showed up again. 

I am blocked with further testing on the banner because it does not shows up for any users in the system if its closed once.

**Comment by catalintomai** [comment]
> Hi @catalintomai
> 
> I am seeing the following issues:
> 
> 1. Header is not changed for 200-user and 400-user advisory. As per the JIRA ticket, it is expected to change the header for these advisory messages as well.
> 2. Banner does not reappear when the nag interval has passed. As per the expectations mentioned in the JIRA ticket,
> 
> `Make banner not reappear upon page refresh (it should only reappear when another advisory is triggered or when the nag interval has passed)`
> 
> Infact, for me even when I removed the warn_metric_number_of_active_users_500 entry from systems table, the banner never showed up again.
> 
> I am blocked with further testing on the banner because it does not shows up for any users in the system if its closed once.

@srkgupta 
#1  - this seems to be an addition to the current design - I'll add the title
#2 - to my understanding the banner should reappear when the nag interval is hit again - but not on refresh anymore(e.g. I set the nag interval to every 2 minutes and for me the banner reappears every 2 minutes if I set the number of active users > 500).

**Comment by srkgupta** [comment]
@catalintomai can you dismiss the banner and then check if it appears again. I was not able to get the banner be displayed again once it was dismissed once.

**Comment by catalintomai** [comment]
> @catalintomai can you dismiss the banner and then check if it appears again. I was not able to get the banner be displayed again once it was dismissed once.

@srkgupta I've sent you a video (longer than a minute) -  I can't seem to see the behavior you describe.

> @catalintomai can you dismiss the banner and then check if it appears again. I was not able to get the banner be displayed again once it was dismissed once.

@srkgupta - sent you a DM with a video - it shows the banner re-appearing after the interval expires.

**Comment by srkgupta** [comment]
Hi @catalintomai 
I did a fresh DB install and it started working fine and I was able to proceed further and test the remaining tests with the banner. I am now seeing the following issue.  Can you please fix the same:

1. When the user clicks on Acknowledge button, the banner now displays the message "Thank you for contacting Mattermost. We will follow up with you soon." However there is no X button when this message is shown. As a result, the user will not be able to dismiss this message ever again until the user reloads the page. Can you please add a Dismiss button to this message 
Screenshot 2020-09-03 at 1 47 39 PM

**Comment by catalintomai** [comment]
@srkgupta - I asked exactly this question of @esadur yesterday =). Eric - should we add the X button back for the ack banner?

**Comment by esadur** [comment]
> @srkgupta - I asked exactly this question of @esadur yesterday =). Eric - should we add the X button back for the ack banner?

@catalintomai yes, we should add the X button for the success banner.

**Review by srkgupta — APPROVED** [review]
Tested the updated PR and all the issues are fixed now and the PR is working fine as expected. Approving the PR.

**Comment by catalintomai** [comment]
@amyblais - manually cherry-picked into 5.27

**Pull request merged**